REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [297]:
import pandas as pd
import numpy as np

In [424]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [10]:
bus.head(6)

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   
5           48                          ART'S CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   
5                 747 IRVING ST   San Francisco    CA       94122  37.764013   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051  
5 -122.465749  +14156657440

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [11]:
bus.shape[0]

6315

#### Question 2b: How many unique business IDs are there?  

In [441]:
bus.business_id.nunique()

6315

In [13]:
print(bus['business_id'])

0          10
1          19
2          24
3          31
4          45
        ...  
6310    90234
6311    90236
6312    90237
6313    90238
6314    90239
Name: business_id, Length: 6315, dtype: int64


#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [14]:
n = 5
bus['name'].value_counts()[:n].index.tolist()

['STARBUCKS COFFEE',
 "PEET'S COFFEE & TEA",
 'MCDONALDS',
 'SAN FRANCISCO SOUP COMPANY',
 'WALGREENS']

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [16]:
bus.dtypes['postal_code']

dtype('O')

In [17]:
bus.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6315 entries, 0 to 6314
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   business_id   6315 non-null   int64  
 1   name          6315 non-null   object 
 2   address       6315 non-null   object 
 3   city          6315 non-null   object 
 4   state         6315 non-null   object 
 5   postal_code   6056 non-null   object 
 6   latitude      3599 non-null   float64
 7   longitude     3599 non-null   float64
 8   phone_number  5696 non-null   object 
dtypes: float64(2), int64(1), object(6)
memory usage: 444.1+ KB


#### Question 3b

What are the unique values of postal_code?

In [18]:
bus.postal_code.unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [19]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [20]:
vb = bus[bus['postal_code'].isin(validZip)]

In [21]:
vb.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5999 entries, 0 to 6314
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   business_id   5999 non-null   int64  
 1   name          5999 non-null   object 
 2   address       5999 non-null   object 
 3   city          5999 non-null   object 
 4   state         5999 non-null   object 
 5   postal_code   5999 non-null   object 
 6   latitude      3516 non-null   float64
 7   longitude     3516 non-null   float64
 8   phone_number  5444 non-null   object 
dtypes: float64(2), int64(1), object(6)
memory usage: 468.7+ KB


In [22]:
vb.shape[0]

5999

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [132]:
vb[pd.isnull(vb['longitude'])]

business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Francisco    CA       94110       NaN        NaN  +14156481034  
210   San Francisco    CA       94133       NaN        NaN  +14153979919  
760   San Francisco    CA       94133       NaN        NaN  +14157715687  
826   San Francisco    CA       94102       NaN        NaN  +14153975397  
1100  San Francisco    CA       94109       NaN        NaN  +14154743482  
...             ...   ...         ...       ...        ...           ...  
6310  San Francisco    CA       94103       NaN        NaN           NaN  
6311  San Francisco    CA       94127       NaN        NaN  +14154945048  
6312  San Francisco    CA       94116       NaN        NaN  +14158606036  
6313  San Francisco    CA       94112       NaN        NaN  +14153331878  
6314  San Francisco    CA       94110       NaN        NaN  +14158881393  

[2483 rows x 9 columns]

In [133]:
vb.longitude.isnull().sum()

2483

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [134]:
df2 = pd.DataFrame().assign(postal_code=vb['postal_code'], longitude=vb['longitude'])

df2

postal_code   longitude
0          94104 -122.403816
1          94109 -122.421547
2          94104 -122.403135
3          94133 -122.419004
4          94110 -122.413641
...          ...         ...
6310       94103         NaN
6311       94127         NaN
6312       94116         NaN
6313       94112         NaN
6314       94110         NaN

[5999 rows x 2 columns]

In [135]:
df2['null_lon'] = df2['longitude'].isnull()*True


In [136]:
df2.rename(
    columns={"longitude":"not_null_long"}
          ,inplace=True)

df2

postal_code  not_null_long  null_lon
0          94104    -122.403816     False
1          94109    -122.421547     False
2          94104    -122.403135     False
3          94133    -122.419004     False
4          94110    -122.413641     False
...          ...            ...       ...
6310       94103            NaN      True
6311       94127            NaN      True
6312       94116            NaN      True
6313       94112            NaN      True
6314       94110            NaN      True

[5999 rows x 3 columns]

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [157]:
df2.groupby(['postal_code'])['not_null_long'].apply(lambda x: x.isnull().sum())

postal_code
94102    196
94103    252
94104     59
94105    100
94107    247
94108     82
94109    146
94110    251
94111    115
94112     59
94114    108
94115     76
94116     35
94117     67
94118    100
94121     55
94122    123
94123     53
94124     88
94127     28
94131     14
94132     65
94133    135
94134     29
Name: not_null_long, dtype: int64

In [164]:
df3 = df2.groupby(['postal_code'])['not_null_long'].apply(lambda x: x.isnull().sum())

print(df3)

postal_code
94102    196
94103    252
94104     59
94105    100
94107    247
94108     82
94109    146
94110    251
94111    115
94112     59
94114    108
94115     76
94116     35
94117     67
94118    100
94121     55
94122    123
94123     53
94124     88
94127     28
94131     14
94132     65
94133    135
94134     29
Name: not_null_long, dtype: int64


In [178]:
df3['sum_null_perzip'] = df2.groupby(['postal_code'])['not_null_long'].apply(lambda x: x.isnull().sum())

print(df3['sum_null_perzip'].nlargest(n=5))


postal_code
94103    252
94110    251
94107    247
94102    196
94109    146
Name: not_null_long, dtype: int64


# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [425]:
ins

business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15430 rows x 4 columns]

In [426]:
ins.count()

business_id    15430
score          15430
date           15430
type           15430
dtype: int64

In [427]:
ins.business_id.value_counts()

5626     8
5487     8
5461     8
5633     8
5471     8
        ..
80931    1
80925    1
71413    1
80897    1
89900    1
Name: business_id, Length: 5730, dtype: int64

In [428]:
ins.business_id.nunique()

5730

#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [429]:
ins.type.value_counts()

routine      15429
complaint        1
Name: type, dtype: int64

In [430]:
#ins2 = ins.copy()

#ins2[ins.groupby('type')['type'].transform('size') > 10]

ins2 = ins[['business_id', 'score', 'date', 'type']]
ins[ins2.replace(ins2.stack().value_counts()).gt(10).all(1)]

business_id  score      date     type
31           66    100  20160902  routine
32           66     98  20160322  routine
33           66    100  20150828  routine
34           66     98  20141119  routine
35           66     91  20140519  routine
36           66     98  20140210  routine
37           67     94  20161019  routine
38           67     87  20160401  routine
39           67     90  20150520  routine
40           67     94  20140205  routine
41           73     92  20160927  routine
43           73     92  20140522  routine
44           76     98  20161005  routine
45           76    100  20160311  routine
46           76    100  20141124  routine
47           76     96  20140612  routine
48           77     96  20161005  routine
49           77     91  20160311  routine
50           77     96  20141124  routine
51           77     94  20140612  routine
52           80     98  20160412  routine
53           80    100  20140520  routine
54           81     96  20150330  routine
55           81     92  20140210  routine
56           88     89  20150824  routine
58           88     78  20140211  routine
63           98     96  20160316  routine
65           98     96  20140324  routine

In [431]:
ins2.type.value_counts()

routine      15429
complaint        1
Name: type, dtype: int64

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [432]:
ins2['year'] = ins2['date'].astype(str).str[0:4]


In [433]:
ins2.head()

business_id  score      date     type  year
0           10     82  20160503  routine  2016
1           10     94  20140729  routine  2014
2           10     92  20140114  routine  2014
3           19     94  20160513  routine  2016
4           19     94  20141110  routine  2014

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [434]:
ins2.year.value_counts()

2016    5840
2014    5629
2015    3923
2013      38
Name: year, dtype: int64

In [435]:
ins3 = ins2.groupby('year').filter(lambda x : (x['year'].count()>=50).any())


In [437]:
ins3.year.value_counts()

2016    5840
2014    5629
2015    3923
Name: year, dtype: int64

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [438]:
ins_2016 = ins3.drop(ins3[ins3.year.isin(["2014", "2015"])].index)


In [439]:
ins_2016.head()

business_id  score      date     type  year
0            10     82  20160503  routine  2016
3            19     94  20160513  routine  2016
6            24     98  20161005  routine  2016
7            24     96  20160311  routine  2016
10           31     84  20160816  routine  2016

In [440]:
ins_2016.year.value_counts()

2016    5840
Name: year, dtype: int64

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [442]:
bus_insp_merge = pd.merge(bus, ins_2016, on='business_id')
bus_insp_merge.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4           31      NORMAN'S ICE CREAM AND FREEZES   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
4           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   

    longitude  phone_number  score      date     type  year  
0 -122.403816  +14154217044     82  20160503  routine  2016  
1 -122.421547  +14157763262     94  20160513  routine  2016  
2 -122.403135  +14156779494     98  20161005  routine  2016  
3 -122.403135  +14156779494     96  20160311  routine  2016  
4 -122.419004           NaN     84  20160816  routine  2016

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [487]:
bus_insp_merge.drop(['latitude', 'longitude', 'phone_number', 'date', 'type', 'year'], axis=1, inplace=True)

In [491]:
bus_insp_merge.nsmallest(20, ['score'])

business_id                         name              address  \
3931        71373      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD    
3917        71232                   UNCLE CAFE       65 WAVERLY PL    
4981        81667                 CRAZY PEPPER   2257 SAN JOSE AVE    
5727        87652                    POKI TIME     2101 LOMBARD ST    
2488        34585             HING WANG BAKERY        339 JUDAH ST    
4481        78328                   GOLDEN WOK   295 B ORIZABA AVE    
457          1843          MA'S DIM SUM & CAFE      1315 POWELL ST    
696          2716                 RED A BAKERY      634 CLEMENT ST    
3415        67330      SAN TUNG RESTAURANT LLC      1031 IRVING ST    
3940        71440  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST    
4298        76615                   TOP'S CAFE  2460 SAN BRUNO AVE    
4824        80591    L & G VIETNAMESE SANDWICH         602 EDDY ST    
5226        83647    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST    
523          2044          JOANIE'S DINER INC.   1329 COLUMBUS AVE    
2927        61161                  JADE GARDEN      3608 BALBOA ST    
3430        67564                        ITHAI         720 POST ST    
5424        85587         PAK NAM THAI CUISINE       655 LARKIN ST    
3904        70996             HONG KONG LOUNGE     5322 GEARY BLVD    
4025        72176                 WOLFES LUNCH        1220 16TH ST    
4885        81106          VALLARTA'S TACO BAR     2022 MISSION ST    

               city state postal_code  score  
3931  San Francisco    CA       94121     52  
3917  San Francisco    CA       94108     55  
4981  San Francisco    CA       94112     55  
5727  San Francisco    CA       94123     55  
2488  San Francisco    CA       94122     56  
4481  San Francisco    CA       94132     56  
457   San Francisco    CA       94133     57  
696   San Francisco    CA       94118     57  
3415  San Francisco    CA       94122     57  
3940  San Francisco    CA       94108     57  
4298  San Francisco    CA       94134     57  
4824  San Francisco    CA       94109     57  
5226  San Francisco    CA       94123     58  
523   San Francisco    CA       94133     59  
2927  San Francisco    CA       94121     59  
3430  San Francisco    CA       94109     59  
5424  San Francisco    CA       94109     59  
3904  San Francisco    CA       94121     60  
4025  San Francisco    CA       94107     60  
4885  San Francisco    CA       94110     60

In [490]:
print(bus_insp_merge.nsmallest(20, ['score']))

      business_id                         name              address  \
3931        71373      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD    
3917        71232                   UNCLE CAFE       65 WAVERLY PL    
4981        81667                 CRAZY PEPPER   2257 SAN JOSE AVE    
5727        87652                    POKI TIME     2101 LOMBARD ST    
2488        34585             HING WANG BAKERY        339 JUDAH ST    
4481        78328                   GOLDEN WOK   295 B ORIZABA AVE    
457          1843          MA'S DIM SUM & CAFE      1315 POWELL ST    
696          2716                 RED A BAKERY      634 CLEMENT ST    
3415        67330      SAN TUNG RESTAURANT LLC      1031 IRVING ST    
3940        71440  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST    
4298        76615                   TOP'S CAFE  2460 SAN BRUNO AVE    
4824        80591    L & G VIETNAMESE SANDWICH         602 EDDY ST    
5226        83647    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST    
523   

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!